source/geh_common/src/geh_common/testing/dataframes/assert_dataframes.py

```diff
@@ -122,6 +123,45 @@ def assert_dataframes_and_schemas(
         )
         raise
 
+    # Check row order if configured
+    if configuration.enforce_row_order:
+        try:
+            assert_row_order(actual, expected)
+        except AssertionError:
+            if not configuration.show_columns_when_actual_and_expected_are_equal:
+                actual, expected = _drop_columns_if_the_same(actual, expected)
+
+            print("ROW ORDER MISMATCH:")  # noqa
+            print("First 10 rows of ACTUAL:")  # noqa
+            actual.limit(10).show(truncate=False)
+            print("First 10 rows of EXPECTED:")  # noqa
+            expected.limit(10).show(truncate=False)
+            raise
+
+
+def assert_row_order(actual: DataFrame, expected: DataFrame) -> None:
+    """Assert that rows in actual dataframe match the exact order of rows in expected dataframe.
+
+    Args:
+        actual: The actual dataframe
+        expected: The expected dataframe
+
+    Raises:
+        AssertionError: If rows do not appear in the same order
+    """
+    # Convert both dataframes to lists of rows for direct comparison
+    actual_rows = actual.collect()
+    expected_rows = expected.collect()
```

Comment:
How many times are the data frames materialized? They are counted, collected, shown, and more. We should probably fix this to ensure fast scenario tests.

What's the best approach? Cache the data frames as the first thing in the assert functions?

Reply:
4 collect(), 10 show(), 8 count()

Reply:
should I make a story for this or just try to fix it now?